File: apps/workbench/src-tauri/src/commands/task_control.rs
//! Explicit typed Workbench -> Run Manager task-control dispatch.

use devbox_applink::{
    CreateHandoff, HandoffError, HandoffStore, OpenRequest, TaskControlAction, TaskControlRequest,
    TASK_CONTROL_HANDOFF_KIND, TASK_CONTROL_SCHEMA_VERSION, TASK_CONTROL_SOURCE_APP,
    TASK_CONTROL_TARGET_APP,
};
use serde::{Deserialize, Serialize};
use std::collections::BTreeSet;

const RUN_MANAGER_SCHEMA_VERSION: u32 = 1;
const WORKSPACE_TASKS_VIEW_KIND: &str = "workspace-tasks";
const TASK_CONTROL_RECEIPTS_VIEW_KIND: &str = "task-control-receipts";
const MAX_WORKSPACE_TASKS: usize = 128;

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct WorkspaceTaskControlItem {
    pub id: String,
    pub label: String,
    pub revision: String,
    pub task_kind: String,
    pub trusted: bool,
    pub shell_trusted: bool,
    pub available: bool,
    pub has_dependencies: bool,
    pub operation_active: bool,
}

#[derive(Debug, Clone, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct TaskControlReceipt {
    pub schema_version: u32,
    pub request_id: String,
    pub task_id: String,
    pub action: TaskControlAction,
    pub status: String,
    pub operation_id: Option<String>,
    pub failure_code: Option<String>,
    pub created_at: i64,
    pub updated_at: i64,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TaskControlDispatch {
    pub request_id: String,
    pub handoff_id: String,
}

fn handoff_store() -> HandoffStore {
    HandoffStore::new(devbox_applink::handoff_root_in(
        &devbox_integration::common_root(),
    ))
}

fn now_ms() -> u64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|duration| duration.as_millis() as u64)
        .unwrap_or(0)
}

#[tauri::command]
pub fn dispatch_workspace_task_control(
    task_id: String,
    action: TaskControlAction,
    expected_revision: String,
) -> Result<TaskControlDispatch, String> {
    // Treat the renderer selection as a hint. Re-read the strict native
    // snapshot before publishing so an injected renderer value cannot become
    // durable handoff/receipt metadata, even though Run Manager also verifies
    // the request independently.
    let controls = list_workspace_task_controls()?;
    authorize_dispatch(&controls, &task_id, action, &expected_revision).map_err(str::to_owned)?;
    let request = TaskControlRequest {
        schema_version: TASK_CONTROL_SCHEMA_VERSION,
        request_id: uuid::Uuid::new_v4().simple().to_string(),
        task_id,
        action,
        expected_revision,
    };
    let payload = request.to_payload().map_err(str::to_owned)?;
    let now = now_ms();
    if now == 0 {
        return Err("task-control-unavailable".to_owned());
    }
    let store = handoff_store();
    let publication = store
        .create_with_publication(
            CreateHandoff {
                kind: TASK_CONTROL_HANDOFF_KIND.to_owned(),
                source_app: TASK_CONTROL_SOURCE_APP.to_owned(),
                target_app: Some(TASK_CONTROL_TARGET_APP.to_owned()),
                payload,
            },
            now,
        )
        .map_err(|_| "task-control-unavailable".to_owned())?;
    let open = OpenRequest {
        target: publication.descriptor.clone().into(),
        from: Some(TASK_CONTROL_SOURCE_APP.to_owned()),
    };
    if devbox_launch::launch_open(TASK_CONTROL_TARGET_APP, &open).is_err() {
        match store.remove_pending(&publication) {
            Ok(()) | Err(HandoffError::Missing) => {}
            Err(_) => return Err("task-control-cleanup-failed".to_owned()),
        }
        return Err("task-control-run-manager-unavailable".to_owned());
    }
    Ok(TaskControlDispatch {
        request_id: request.request_id,
        handoff_id: publication.descriptor.id,
    })
}

fn authorize_dispatch(
    controls: &[WorkspaceTaskControlItem],
    task_id: &str,
    action: TaskControlAction,
    expected_revision: &str,
) -> Result<(), &'static str> {
    let task = controls
        .iter()
        .find(|task| task.id == task_id)
        .ok_or("task-control-task-not-found")?;
    if task.revision != expected_revision {
        return Err("task-control-source-changed");
    }
    match action {
        TaskControlAction::Start if task.operation_active => Err("workspace-task-operation-active"),
        TaskControlAction::Start if !task.available => Err("workspace-task-unavailable"),
        TaskControlAction::Start if !task.trusted => Err("workspace-task-source-untrusted"),
        TaskControlAction::Start if task.task_kind == "shell" && !task.shell_trusted => {
            Err("workspace-task-shell-untrusted")
        }
        TaskControlAction::Stop if !task.operation_active => {
            Err("task-control-operation-not-active")
        }
        TaskControlAction::Start | TaskControlAction::Stop => Ok(()),
    }
}

#[tauri::command]
pub fn list_workspace_task_controls() -> Result<Vec<WorkspaceTaskControlItem>, String> {
    let envelope = devbox_integration::read_named_view_snapshot_in(
        &devbox_integration::integration_root(),
        "run-manager",
        RUN_MANAGER_SCHEMA_VERSION,
        WORKSPACE_TASKS_VIEW_KIND,
    )
    .map_err(|_| "task-control-snapshot-unavailable".to_owned())?
    .ok_or_else(|| "task-control-snapshot-missing".to_owned())?;
    let views = envelope
        .views()
        .map_err(|_| "task-control-snapshot-invalid".to_owned())?;
    let view = views
        .get(WORKSPACE_TASKS_VIEW_KIND)
        .ok_or_else(|| "task-control-snapshot-invalid".to_owned())?;
    if view.schema_version != 1 || view.entries.len() > MAX_WORKSPACE_TASKS {
        return Err("task-control-snapshot-invalid".to_owned());
    }
    let mut ids = BTreeSet::new();
    let mut tasks = Vec::with_capacity(view.entries.len());
    for value in &view.entries {
        let task: WorkspaceTaskControlItem = serde_json::from_value(value.clone())
            .map_err(|_| "task-control-snapshot-invalid".to_owned())?;
        let probe = TaskControlRequest {
            schema_version: TASK_CONTROL_SCHEMA_VERSION,
            request_id: "a".repeat(32),
            task_id: task.id.clone(),
            action: TaskControlAction::Start,
            expected_revision: task.revision.clone(),
        };
        if probe.validate().is_err()
            || !ids.insert(task.id.clone())
            || !matches!(task.task_kind.as_str(), "process" | "shell")
            || task.label.is_empty()
            || task.label.len() > 256
            || task.label.chars().any(char::is_control)
        {
            return Err("task-control-snapshot-invalid".to_owned());
        }
        tasks.push(task);
    }
    tasks.sort_by(|left, right| left.label.cmp(&right.label).then(left.id.cmp(&right.id)));
    Ok(tasks)
}

#[tauri::command]
pub fn get_workspace_task_control_receipt(
    request_id: String,
) -> Result<Option<TaskControlReceipt>, String> {
    let request_probe = TaskControlRequest {
        schema_version: TASK_CONTROL_SCHEMA_VERSION,
        request_id: request_id.clone(),
        task_id: "probe".to_owned(),
        action: TaskControlAction::Start,
        expected_revision: "a".repeat(64),
    };
    request_probe
        .validate()
        .map_err(|_| "task-control-request-invalid".to_owned())?;
    let Some(envelope) = devbox_integration::read_named_view_snapshot_in(
        &devbox_integration::integration_root(),
        "run-manager",
        RUN_MANAGER_SCHEMA_VERSION,
        TASK_CONTROL_RECEIPTS_VIEW_KIND,
    )
    .map_err(|_| "task-control-receipt-unavailable".to_owned())?
    else {
        return Ok(None);
    };
    let views = envelope
        .views()
        .map_err(|_| "task-control-receipt-invalid".to_owned())?;
    let view = views
        .get(TASK_CONTROL_RECEIPTS_VIEW_KIND)
        .ok_or_else(|| "task-control-receipt-invalid".to_owned())?;
    if view.schema_version != 1 || view.entries.len() > 100 {
        return Err("task-control-receipt-invalid".to_owned());
    }
    let mut found = None;
    let mut ids = BTreeSet::new();
    for value in &view.entries {
        let receipt: TaskControlReceipt = serde_json::from_value(value.clone())
            .map_err(|_| "task-control-receipt-invalid".to_owned())?;
        if !valid_receipt(&receipt) || !ids.insert(receipt.request_id.clone()) {
            return Err("task-control-receipt-invalid".to_owned());
        }
        if receipt.request_id == request_id {
            found = Some(receipt);
        }
    }
    Ok(found)
}

fn valid_receipt(receipt: &TaskControlReceipt) -> bool {
    let request_probe = TaskControlRequest {
        schema_version: receipt.schema_version,
        request_id: receipt.request_id.clone(),
        task_id: receipt.task_id.clone(),
        action: receipt.action,
        // Receipts deliberately do not disclose the expected source revision.
        // A fixed valid digest lets the shared contract validate every other
        // request field without reconstructing that private input.
        expected_revision: "a".repeat(64),
    };
    if request_probe.validate().is_err()
        || receipt.created_at <= 0
        || receipt.updated_at < receipt.created_at
    {
        return false;
    }

    let operation_valid = receipt.operation_id.as_deref().is_some_and(|value| {
        uuid::Uuid::parse_str(value).is_ok_and(|parsed| parsed.to_string() == value)
    });
    let failure_valid = receipt.failure_code.as_deref().is_some_and(|value| {
        !value.is_empty()
            && value.len() <= 64
            && value
                .bytes()
                .all(|byte| byte.is_ascii_lowercase() || byte.is_ascii_digit() || byte == b'-')
    });

    match receipt.status.as_str() {
        "accepted" => receipt.operation_id.is_none() && receipt.failure_code.is_none(),
        "rejected" | "failed" => receipt.operation_id.is_none() && failure_valid,
        "started" | "stopped" => operation_valid && receipt.failure_code.is_none(),
        _ => false,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dispatch_shape_has_only_opaque_correlators() {
        let dispatch = TaskControlDispatch {
            request_id: "a".repeat(32),
            handoff_id: "b".repeat(32),
        };
        let value = serde_json::to_value(dispatch).unwrap();
        assert_eq!(
            value
                .as_object()
                .unwrap()
                .keys()
                .cloned()
                .collect::<Vec<_>>(),
            vec!["handoffId", "requestId"]
        );
    }

    #[test]
    fn receipt_validation_rejects_untrusted_snapshot_fields() {
        let mut receipt = TaskControlReceipt {
            schema_version: TASK_CONTROL_SCHEMA_VERSION,
            request_id: "a".repeat(32),
            task_id: "task:build".to_owned(),
            action: TaskControlAction::Start,
            status: "started".to_owned(),
            operation_id: Some("550e8400-e29b-41d4-a716-446655440000".to_owned()),
            failure_code: None,
            created_at: 1,
            updated_at: 2,
        };
        assert!(valid_receipt(&receipt));

        receipt.operation_id = Some("../operation".to_owned());
        assert!(!valid_receipt(&receipt));
        receipt.operation_id = None;
        receipt.status = "failed".to_owned();
        receipt.failure_code = Some("FAILED_WITH_RAW_TEXT".to_owned());
        assert!(!valid_receipt(&receipt));
        receipt.failure_code = Some("task-control-interrupted".to_owned());
        assert!(valid_receipt(&receipt));
        receipt.updated_at = 0;
        assert!(!valid_receipt(&receipt));
    }

    #[test]
    fn dispatch_authorization_rechecks_the_exact_native_snapshot_state() {
        let task = WorkspaceTaskControlItem {
            id: "550e8400-e29b-41d4-a716-446655440000".to_owned(),
            label: "Build".to_owned(),
            revision: "a".repeat(64),
            task_kind: "process".to_owned(),
            trusted: true,
            shell_trusted: false,
            available: true,
            has_dependencies: false,
            operation_active: false,
        };
        assert_eq!(
            authorize_dispatch(
                std::slice::from_ref(&task),
                &task.id,
                TaskControlAction::Start,
                &task.revision,
            ),
            Ok(())
        );
        assert_eq!(
            authorize_dispatch(
                std::slice::from_ref(&task),
                "arbitrary-renderer-value",
                TaskControlAction::Start,
                &task.revision,
            ),
            Err("task-control-task-not-found")
        );
        assert_eq!(
            authorize_dispatch(
                std::slice::from_ref(&task),
                &task.id,
                TaskControlAction::Start,
                &"b".repeat(64),
            ),
            Err("task-control-source-changed")
        );
        assert_eq!(
            authorize_dispatch(
                std::slice::from_ref(&task),
                &task.id,
                TaskControlAction::Stop,
                &task.revision,
            ),
            Err("task-control-operation-not-active")
        );
    }
}

File: packages/workspace-features/src/overview/lib/taskControl.ts
import type {
  WorkspaceTaskControl,
  WorkspaceTaskControlReceipt,
  WorkspaceTaskControlReceiptStatus,
} from "../api";

/**
 * Task-control errors are intentionally translated from a closed set. Native
 * error text can contain paths or other local details and must never become UI
 * output in Workbench.
 */
const FIXED_TASK_CONTROL_ERRORS: Record<string, string> = {
  "task-control-dispatch-invalid": "Run Manager task 요청 형식이 올바르지 않습니다.",
  "task-control-snapshot-unavailable": "Run Manager task 상태를 읽을 수 없습니다. 잠시 후 다시 시도하세요.",
  "task-control-snapshot-missing": "Run Manager task snapshot이 없습니다. Run Manager를 열고 task를 새로고침하세요.",
  "task-control-snapshot-invalid": "Run Manager task snapshot을 안전하게 읽을 수 없습니다.",
  "task-control-receipt-unavailable": "Run Manager task 확인 결과를 읽을 수 없습니다.",
  "task-control-receipt-invalid": "Run Manager task 확인 결과가 올바르지 않습니다.",
  "task-control-request-invalid": "Run Manager task 요청 값이 올바르지 않습니다.",
  "task-control-run-manager-unavailable": "Run Manager를 열 수 없어 task 요청을 전달하지 못했습니다.",
  "task-control-cleanup-failed": "전달하지 못한 Run Manager task 요청을 정리하지 못했습니다.",
  "task-control-task-not-found": "Run Manager에서 이 workspace task를 찾지 못했습니다. 목록을 새로고침하세요.",
  "task-control-source-changed": "task snapshot 이후 원본 revision이 변경되었습니다. 목록을 새로고침하고 다시 확인하세요.",
  "task-control-operation-not-active": "현재 실행 중인 workspace task operation이 없습니다.",
  "task-control-request-replayed": "이미 처리된 task 요청입니다. 최신 상태를 새로고침하세요.",
  "task-control-user-rejected": "Run Manager에서 task 요청을 거절했습니다.",
  "task-control-interrupted": "Run Manager가 종료되어 처리 중이던 task 요청이 중단되었습니다.",
  "task-control-claim-failed": "Run Manager가 task 요청의 소유권을 확정하지 못했습니다.",
  "task-control-receipt-storage": "Run Manager가 task 확인 결과를 저장하지 못했습니다.",
  "workspace-task-source-untrusted": "이 workspace task의 소스가 아직 승인되지 않았습니다. Run Manager에서 현재 revision을 승인하세요.",
  "workspace-task-shell-untrusted": "이 셸 task의 실행이 아직 승인되지 않았습니다. Run Manager에서 셸 실행을 별도로 승인하세요.",
  "workspace-task-unavailable": "이 workspace task는 현재 사용할 수 없습니다. Run Manager에서 원본을 다시 확인하세요.",
  "workspace-task-source-changed": "원본 tasks.json이 변경되어 승인이 무효화되었습니다. Run Manager에서 다시 확인하세요.",
  "workspace-task-dependency-selection-incomplete": "선택한 task의 선행 dependency가 빠졌습니다. Run Manager에서 dependency를 함께 선택하세요.",
  "workspace-task-dependency-unavailable": "선행 dependency를 사용할 수 없어 task를 실행할 수 없습니다.",
  "workspace-task-dependency-cycle": "task dependency에 순환 참조가 있어 실행할 수 없습니다.",
  "workspace-task-orchestration-required": "이 workspace task는 dependency가 있어 Run Manager orchestration이 필요합니다.",
  "workspace-task-operation-active": "이 workspace task에는 이미 실행 중인 Run Manager operation이 있습니다.",
  "workspace-task-operation-not-found": "workspace task operation을 찾지 못했습니다. task 목록을 새로고침하세요.",
  "workspace-task-operation-state-changed": "workspace task operation 상태가 바뀌어 요청을 완료하지 못했습니다.",
  "workspace-task-operation-cancelled": "workspace task operation이 취소되었습니다.",
  "workspace-task-operation-stop-failed": "workspace task operation을 안전하게 중지하지 못했습니다.",
  "workspace-task-operation-stop-timeout": "workspace task operation 중지가 제한 시간 안에 끝나지 않았습니다.",
  "workspace-task-operation-ownership-changed": "workspace task operation의 실행 소유권이 바뀌어 중지할 수 없습니다.",
  "workspace-task-start-failed": "workspace task를 시작하지 못했습니다.",
  "workspace-task-run-failed": "workspace task 실행이 실패했습니다.",
  "workspace-task-run-cancelled": "workspace task 실행이 취소되었습니다.",
  "workspace-task-dependency-failed": "선행 dependency가 실패해 operation을 진행하지 못했습니다.",
};

export function taskControlErrorMessage(cause: unknown): string {
  const value = cause instanceof Error ? cause.message : String(cause);
  const code = value.trim();
  return FIXED_TASK_CONTROL_ERRORS[code]
    ?? "Run Manager task 요청을 완료하지 못했습니다.";
}

export function canStartWorkspaceTask(task: WorkspaceTaskControl): boolean {
  return !task.operationActive
    && task.trusted
    && task.available
    && (task.taskKind === "process" || task.shellTrusted);
}

export function canStopWorkspaceTask(task: WorkspaceTaskControl): boolean {
  return task.operationActive;
}

export function isTerminalTaskControlReceipt(
  status: WorkspaceTaskControlReceiptStatus,
): boolean {
  return status === "rejected"
    || status === "started"
    || status === "stopped"
    || status === "failed";
}

export function taskControlReceiptMessage(receipt: WorkspaceTaskControlReceipt): string {
  if (receipt.status === "started") return "Run Manager가 시작했습니다.";
  if (receipt.status === "stopped") return "Run Manager가 중지했습니다.";
  if (receipt.status === "rejected" || receipt.status === "failed") {
    return taskControlErrorMessage(receipt.failureCode ?? receipt.status);
  }
  return "Run Manager 창의 확인을 기다리는 중…";
}
